internal/services/authorization/role_definition_data_source.go

```diff
@@ -4,178 +4,251 @@
-	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonids"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/authorization/2022-05-01-preview/roledefinitions"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
-	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
 )
 
-func dataSourceArmRoleDefinition() *pluginsdk.Resource {
-	return &pluginsdk.Resource{
-		Read: dataSourceArmRoleDefinitionRead,
+type ArmRoleDefinitionDataSource struct{}
 
-		Timeouts: &pluginsdk.ResourceTimeout{
-			Read: pluginsdk.DefaultTimeout(5 * time.Minute),
-		},
+var _ sdk.DataSource = ArmRoleDefinitionDataSource{}
 
-		Schema: map[string]*pluginsdk.Schema{
-			"name": {
-				Type:          pluginsdk.TypeString,
-				Optional:      true,
-				Computed:      true,
-				ConflictsWith: []string{"role_definition_id"},
-			},
+type ArmRoleDefinitionDataSourceModel struct {
+	Name             string                      `tfschema:"name"`
+	RoleDefinitionId string                      `tfschema:"role_definition_id"`
+	Scope            string                      `tfschema:"scope"`
+	Description      string                      `tfschema:"description"`
+	Type             string                      `tfschema:"type"`
+	Permissions      []PermissionDataSourceModel `tfschema:"permissions"`
+	AssignableScopes []string                    `tfschema:"assignable_scopes"`
+}
+
+type PermissionDataSourceModel struct {
+	Actions        []string `tfschema:"actions"`
+	NotActions     []string `tfschema:"not_actions"`
+	DataActions    []string `tfschema:"data_actions"`
+	NotDataActions []string `tfschema:"not_data_actions"`
+}
 
-			"role_definition_id": {
-				Type:          pluginsdk.TypeString,
-				Optional:      true,
-				Computed:      true,
-				ConflictsWith: []string{"name"},
-				ValidateFunc:  validation.Any(validation.IsUUID, validation.StringIsEmpty),
+func (a ArmRoleDefinitionDataSource) Arguments() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			Computed: true,
+			ExactlyOneOf: []string{
+				"name",
+				"role_definition_id",
 			},
+		},
 
-			"scope": {
-				Type:     pluginsdk.TypeString,
-				Optional: true,
+		"role_definition_id": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			Computed: true,
+			ExactlyOneOf: []string{
+				"name",
+				"role_definition_id",
 			},
+			ValidateFunc: validation.Any(validation.IsUUID, validation.StringIsEmpty),
+		},
 
-			// Computed
+		"scope": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
```

Comment:
Shall we add the `commonids.ValidateScopeID`, in which case the `scope` won't be empty string?

Reply:
it's a data source, the user could specify it as a filter or not.